回调函数嵌套回调函数  一直嵌套  就叫回调地狱

比如 读文件时 需要先读一个文件再读一个。。 按循序读就需要在外层读文件函数的回调函数中嵌套另一个读文件的函数，而再读的话 就要在再一个回调函数内部嵌套再另一个读文件的函数

Promise是一个构造函数（把嵌套变成串联）  console.dir(Promise)



其中有两个方法  

- resolve：成功回调函数
- reject：失败回调函数

在prototype上有个.then方法

- Promise的构造函数的实例都可以访问该方法

Promise的实例本身是异步的

- 异步就只有两种状态（可以返回结果 怎么调用两个方法通过.then()）
  - 成功：调用resolve返回结果
  - 失败：调用reject返回结果
- 由于是异步的所以内部无法使用return返回结果到外面



####  使用

```
//为什么放到function中因为放到外面就会立即执行就不叫异步了
function getFile(fpath){
	const fs = require('fs');
    //只有new就会执行这个function
    var promise = new Promise(function(resolve,reject){
        fs.readFile('path',"utf-8",(err,dataStr)=>{
            if(err) return reject(err);
            resolve(dataStr);
        })
    })
    //为什么return因为resolve和reject方法不知从哪来怎么调用return出去让用户自己设置方法的逻辑
    return promise;
}
var p = getFile("./xxx.txt");
//这里自定义成功回调和失败回调
p.then(function(data){
	console.log(data+"ok")
},function(err){
	console.log(err.message);
})


以下是上面的简写

function getFile(fpath){
	const fs = require('fs');
    //只有new就会执行这个function
    return new Promise(function(resolve,reject){
        fs.readFile('path',"utf-8",(err,dataStr)=>{
            if(err) return reject(err);
            resolve(dataStr);
        })
    })
}
getFile("./xxx.txt").then(function(data){
	console.log(data+"ok")
},function(err){
	console.log(err.message);
})


下面开始嵌套读文件   串联方式

//成功的函数必须传  失败的不一定传
getFile('./1.txt').then(function(data){
	//这里拿到文件1的数据
	console.log(data);
	
	//开始读文件2  并把文件2读的结果return出去
	return getFile("./2.txt")
})/*这个.then是上面return的文件2的结果*/.then(function(data){
	//这里拿到文件2的数据
	console.log(data);
	
	//开始读文件3  并把文件3读的结果return出去
	return getFile("./3.txt")
})/*这里拿到文件3的结果*/.then(function(data){
	console.log(data)
})/*如果上面的任何一个.then执行失败就会马上停止进入.catch中去*/.catch(function(err){
	console.log(err.message);
})

```



###  jquery的Promise

```
//这就是jQuery的Promise因为$.ajax返回的就是一个Promise
$.ajax({
	url:"xxx",
	type:get,
	dataType:"json"
}).then(function(data){
	console.log(data)
})
```

